feat(from-event): observables ready to use from the beginning

tonivj5 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Observables are setted in `AfterContentInit` hook.

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

## What is the new behavior?
It allows to compose observables as component's properties and reduces boilerplate (no needed for `AfterContentInit` hook). For example, [`CounterComponent`](https://github.com/tonivj5/from-event/blob/feat/subject/src/app/counter/counter.component.ts).

The observable is ready to use from the beginning. If the content isn't initialized yet, it returns a subject, which will subscribe to the `fromEvent` once it's created. If the observable is resubscribed, the subject will be completed and removed, and the observable returned will be the `fromEvent` observable instead of the subject.

I haven't added tests, because I don't know if it looks ok for you or not :sweat_smile:. If you like this changes, I'll add them :+1:
## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

NetanelBasal commented:
Thanks for the PR. Allow me giving you some cases to test:

1. Subscribe manually in `ngAfterViewInit`: (Note that in this case the `setter `will be called before the `getter`)

```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button')
  click$: Observable<MouseEvent>;

  ngAfterViewInit() {
    this.click$.pipe(untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

2. Subscribe manually in `ngOnInit`:

```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button', { static: true })
  click$: Observable<MouseEvent>;

  ngOnInit() {
    this.click$.pipe(untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

3. Subscribe multiple times:
```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button', { static: true })
  click$: Observable<MouseEvent>;

  ngOnInit() {
    this.click$.pipe(take(2), untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );

    // This stream should be independent of the stream above
    this.click$.pipe(untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

4. I'm not sure why you've added the `destroy` functionality. If you intended to unsubscribe automatically, we can't:

```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button', { static: true })
  click$: Observable<MouseEvent>;

  ngOnInit() {
    // Think what will happen here when the component is destroyed:
    this.click$.pipe(switchMap(() => interval(1000))).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

tonivj5 commented:
Thanks for your feedback, @NetanelBasal! They're good test-cases, and I will add some clarifications.

> 1. Subscribe manually in ngAfterViewInit: (Note that in this case the setter will be called before the getter)

It works as before, nothing changes in this case. Because `initIfNeeded` (I think it needs a better name :laughing:) only inits the `Subject` if the event isn't initialized. All of tests that depends on this behavior are green.

> 3. Subscribe multiple times

Good catch! It produces a bug. I have fixed it :+1:

> 4. I'm not sure why you've added the destroy functionality. If you intended to unsubscribe automatically, we can't

That wasn't the intended of the `destroyToken`. I added it because I wanted to unsubscribe from the `fromEvent` observable once the `Subject` returned is unsubscribed. Anyway, I have noticed that it should be done in other way, because it produces the fail of `3. Subscribe multiple times`. I have improved it, removing the `destroyToken` in the path.

I'll add all of these test-cases. Thanks :smile:

tonivj5 commented:
@NetanelBasal, I have added some tests, I think it cover all your suggestions.

1. This work as before. I haven't added any new test
```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button')
  click$: Observable<MouseEvent>;

  ngAfterViewInit() {
    this.click$.pipe(untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

2. If `static` is true, it (should) work as before. I have covered this case and if `static` is false
```ts
export class TestComponent {
  @FromEvent('click')
  @ViewChild('button', { static: true })
  click$: Observable<MouseEvent>;

  ngOnInit() {
    this.click$.pipe(untilDestroyed(this)).subscribe(
      observer('TestComponent: button$')
    );
  }
}
```

NetanelBasal commented on projects/ngneat/from-event/src/lib/from-event.ts:
```diff
@@ -1,16 +1,26 @@
+          event$.subscribe(this[eventToken]);
+        }
+
+        this[eventToken] = event$;
```
Why we're overriding the original observable? ( the subject )

NetanelBasal replied:
Ok, got it.

tonivj5 replied:
Sorry for the delay :sweat_smile:

We override the original subject, to remove that intermediary, and return the original `fromEvent` in new subscriptions.

NetanelBasal commented:
Few notes before we merge it:
1. You can use `share` instead of `publish` and `refCount`
2. Inline `initIfNeeded` inside the `getter`.
3. Add the `counter` example to the README file.
4. Add yourself as a contributor. 

Thanks!

tonivj5 commented on projects/ngneat/from-event/src/lib/from-event.ts:
```diff
@@ -1,16 +1,28 @@
+        const event$ = fromEvent(elementRef.nativeElement, event, eventOptions);
+
+        if (this[eventToken]) {
+          event$.pipe(takeUntil(this[destroyToken])).subscribe(this[eventToken]);
```
`event$.subscribe(this[eventToken]);` this line may cause a memory leak. In `initIfNeeded` we are holding the subscriptions to the `Subject`, and once there's 0 subscriptors, we complete and remove the subject. But the subscription to the source event (`fromEvent`) follows existing, and pointing to the removed subject (subject which does nothing). It's neccessary to complete the subscription to remove the reference completetly. So, we need the `destroyToken` to force subscription to `fromEvent` to end.

tonivj5 commented:
> 2. Inline `initIfNeeded` inside the `getter`.

Should I inline the `initIfNeeded`'s content inside of both `getters` (`from-event.ts` and `from-events.ts`)? (just to confirm :+1:)

The rest of notes have been done :smile:

Thanks for your time :1st_place_medal:

NetanelBasal commented:
I forgot the other usage. You can keep it.

NetanelBasal commented on README.md:
````diff
@@ -85,6 +85,42 @@ class HostComponent implements AfterViewInit, OnDestroy {
 }
 ```
 
+You can use it from the beginning, without the need of implement the `AfterViewInit` hook.
````
```suggestion
You are not limited to use it only inside `AfterViewInit` or `AfterContentInit`:
```

NetanelBasal commented on README.md:
```diff
@@ -85,6 +85,42 @@ class HostComponent implements AfterViewInit, OnDestroy {
+  @ViewChild('reset')
+  reset$: Observable<MouseEvent>;
+
+  count$ = merge(this.plus$.pipe(mapTo(1)), this.minus$.pipe(mapTo(-1))).pipe(
```
```ts
  count$ = merge(
    this.plus$.pipe(mapTo(1)), 
    this.minus$.pipe(mapTo(-1))
  ).pipe(
    startWith(0),
    scan((acc, curr) => acc + curr)
  );
```

tonivj5 replied:
I formatted as you, but prettier didn't like it :laughing:

NetanelBasal commented on projects/ngneat/from-event/src/lib/init-if-needed.ts:
```diff
@@ -0,0 +1,21 @@
+        event$.complete();
+        event$ = null;
+
+        that[destroyToken].next();
```
Are you sure? Every subscription uses the same subject.

tonivj5 replied:
This block (`finalize`) is executed only when [all subscriptions are completed/unsubscribed](https://stackblitz.com/edit/typescript-zxgi7m?file=index.ts). Once it happens, the subscription to the `fromEvent` and the subjects (`destroyToken` and `eventToken`) aren't needed anymore.

I added a [test](https://github.com/ngneat/from-event/pull/3/files/0d2302f5305adf5b452a94c114f1829d22fd0476#diff-9497f53b33585e43d9421be10d14e60aR58-R64) which should cover this behavior. One subscription doesn't affect to another.

NetanelBasal commented:
Released. Great work!

NetanelBasal commented:
Btw If you want to create a new open source for ngneat, I have something in mind. Let me know.

tonivj5 commented:
Oh, sure! I'm open :smile:

NetanelBasal commented:
A lot of people want to make a library based on this article: https://netbasal.com/make-your-angular-forms-error-messages-magically-appear-1e32350b7fa5

Are you in?

tonivj5 commented:
It's cool! Yeah, I'm in :+1:

NetanelBasal commented:
You can start. Let me know when you have something you want to push and I will create the repo and give you permissions. Use our ngneat/lib project to create it. Call it @ngneat/magic-form-errors.

NetanelBasal commented:
We can chat on Gitter of you need me.

tonivj5 commented:
Perfect :+1:
